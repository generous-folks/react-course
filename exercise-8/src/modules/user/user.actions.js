import { signIn, signOut } from '../../utils/api.utils';
import { getUser } from './user.selectors';

export const LOGIN = 'user/LOGIN';
export const LOGOUT = 'user/LOGOUT';

export const login = (email, password) => async dispatch => {
  try {
    const user = await signIn(email, password);

    localStorage.setItem('user', JSON.stringify(user));

    return dispatch({ type: LOGIN, user });
  } catch (error) {
    dispatch({ type: LOGIN, error });
  }
};

export const logout = () => async (dispatch, getState) => {
  try {
    const user = getUser(getState());
    if (!user) return;

    localStorage.removeItem('user');

    await signOut(user);

    return dispatch({ type: LOGOUT, user });
  } catch (error) {
    dispatch({ type: LOGOUT, error });
  }
};
