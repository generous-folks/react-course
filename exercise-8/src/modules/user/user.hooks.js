export const usePersistedUser = () => {
  return localStorage.getItem('user');
};
