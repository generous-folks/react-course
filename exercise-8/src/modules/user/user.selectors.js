export const isUserConnected = ({ user }) => !!user;
