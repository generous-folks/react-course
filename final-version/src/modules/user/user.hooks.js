export const usePersistedUser = () => {
  // You would normally validate the user token here
  // and set a new one in case it is not valid anymore
  return localStorage.getItem('user');
};
