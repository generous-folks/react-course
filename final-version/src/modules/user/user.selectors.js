export const isUserConnected = ({ user }) => !!user;
export const getUser = ({ user }) => user;
