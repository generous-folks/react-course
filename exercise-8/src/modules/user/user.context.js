import React from 'react';

import { userReducer, initialState } from './user.reducer';

import { dispatchThunk } from '../../utils/context.utils';
import { CHILDREN_PROP_TYPES } from '../../constants/proptypes.constants';
import { useStoredUser } from './user.hooks';

const UserStateContext = React.createContext();
const UserDispatchContext = React.createContext();

const UserProvider = ({ children }) => {
  const user = useStoredUser();
  const updatedState = user && { user };
  const [state, dispatch] = React.useReducer(userReducer, updatedState || initialState);
  const getState = React.useCallback(() => state, [state]);
  return (
    <UserStateContext.Provider value={state}>
      <UserDispatchContext.Provider value={dispatchThunk(dispatch, getState)}>
        {children}
      </UserDispatchContext.Provider>
    </UserStateContext.Provider>
  );
};

UserProvider.propTypes = {
  children: CHILDREN_PROP_TYPES,
};

function useUserState() {
  const context = React.useContext(UserStateContext);
  if (context === undefined) {
    throw new Error('useUserState must be used within a UserProvider');
  }
  return context;
}

function useUserDispatch() {
  const context = React.useContext(UserDispatchContext);
  if (context === undefined) {
    throw new Error('useUserDispatch must be used within a UserProvider');
  }
  return context;
}

function useUser() {
  return [useUserState(), useUserDispatch()];
}

export { UserProvider, useUser, useUserState, useUserDispatch };
