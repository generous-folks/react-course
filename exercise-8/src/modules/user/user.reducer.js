import { LOGIN, LOGOUT } from './user.actions';

export const initialState = {
  user: null,
};

export const userReducer = (state, action) => {
  if (action.error) {
    return { ...state, error: action.error };
  }

  switch (action.type) {
    case LOGIN: {
      return { ...state, user: action.user };
    }

    case LOGOUT: {
      return { ...state, user: null };
    }

    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};
