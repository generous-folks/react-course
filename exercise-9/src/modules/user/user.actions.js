import { signIn, signOut } from '../../utils/api.utils';
import { getUser } from './user.selectors';

export const LOGIN = 'user/LOGIN';
export const LOGOUT = 'user/LOGOUT';

const encryptUserCredentials = (...args) => [...args];

export const login = (email, password) => async dispatch => {
  try {
    const encryptedUser = encryptUserCredentials(email, password);
    const user = await signIn(encryptedUser);

    localStorage.setItem('user', JSON.stringify(user));

    return dispatch({ type: LOGIN, user });
  } catch (error) {
    dispatch({ type: LOGIN, error });
  }
};

export const logout = () => async (dispatch, getState) => {
  try {
    const user = getUser(getState());
    if (!user) return;

    localStorage.removeItem('user');

    await signOut(user);

    return dispatch({ type: LOGOUT, user });
  } catch (error) {
    dispatch({ type: LOGOUT, error });
  }
};
