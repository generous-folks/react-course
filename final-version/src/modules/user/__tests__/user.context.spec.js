import React from 'react';

import { shallow } from 'enzyme';

import { useUser, useUserState, useUserDispatch, UserProvider } from '../user.context';

describe('user.context', () => {
  describe('<UserProvider />', () => {
    it('should render nested providers', () => {
      expect(
        shallow(
          <UserProvider>
            <div />
          </UserProvider>,
        ),
      ).toMatchSnapshot();
    });

    it('should have property value set', () => {
      expect(
        shallow(
          <UserProvider>
            <div />
          </UserProvider>,
        )
          .find('ContextProvider')
          .first()
          .prop('value'),
      ).toMatchObject({ user: null });
    });
  });

  describe('useUserDispatch', () => {
    it('should be defined', () => {
      expect(typeof useUserDispatch).toBe('function');
    });
  });

  describe('useUserState', () => {
    it('should be defined', () => {
      expect(typeof useUserState).toBe('function');
    });
  });

  describe('useUser', () => {
    it('should be defined', () => {
      expect(typeof useUser).toBe('function');
    });
  });
});
