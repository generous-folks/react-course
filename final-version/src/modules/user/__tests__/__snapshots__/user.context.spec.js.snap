// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user.context <UserProvider /> should render nested providers 1`] = `
<ContextProvider
  value={
    Object {
      "user": null,
    }
  }
>
  <ContextProvider
    value={[Function]}
  >
    <div />
  </ContextProvider>
</ContextProvider>
`;
